### DISCUSSION TITLE
Updating PATH in current VSCode session

### DISCUSSION THREAD

**Discussion by msorens** [microsoft/vscode-discussions#516] [Extension Development QnA] [answered]
There must be a straightforward way to solve this but I have had no luck finding it. Here is my scenario:

1. User invokes my `foo: init` command from the command palette.
2. My extension checks if a custom command-line utility (let's call it `fred`) is installed.
3. If not, ask the user whether to install or not. Let's assume the user says to proceed.
4. Download the `fred` binary.
5. Copy to subdir of LOCALAPPDATA and set perms.
6. Update the user search path with a spawned PowerShell subshell:
   `[Environment]::SetEnvironmentVariable("PATH", updatedPath, [EnvironmentVariableTarget]::User)`

That all works fine. If at this point, I close and re-open VSCode, then I can run another command from the command palette (`foo: validate`) that invokes `fred` and it works. (So the PATH update was made persistent.)
But if I do **_not_** restart VSCode, the environment of VSCode has not yet been informed of `updatedPath`, so it fails to find `fred`.

How can I avoid the VSCode restart and refresh the VSCode environment programmatically? If that is not possible, all I really need is that when I use `spawn` to start a subshell, I can somehow inform that subshell about `updatedPath`.

**Comment by bwateratmsft** [comment, marked as answer]
It will depend on whether you are invoking `fred` in the background (i.e. via [`child_process.spawn`](https://nodejs.org/api/child_process.html#child_processspawncommand-args-options) or similar), or if you are executing it as a task, or in a terminal window within VSCode.

## For background processes:

We're doing essentially exactly this with the Azure Developer CLI extension. Here's the relevant code:
https://github.com/Azure/azure-dev/blob/35abe10a19adf806a9b94cf08965380ed2ad549b/ext/vscode/src/utils/azureDevCli.ts#L105-L115

We have some extra guardrails in place that may or may not apply for your scenario.
1. `isWindows()` to do this only on Windows. On Unix, the AZD CLI goes to `/usr/local/bin`, which will (essentially) always already be on the `PATH`.
1. The env var `AZURE_DEV_CLI_PATH` must be unset (used to override the exe path if desired)
1. "Azure Dev CLI" must not already be in the `PATH`
1. The user actually tried to do the install _within this session_. Consequently, other windows will not benefit, and if the user installs by means outside of VSCode, it will not be detected.

Ultimately, it's just adding something to the end of the existing `PATH` variable as needed, before calling `spawn`. An easier-to-use approach would be to always add an entry if it's not already present, i.e., only using guard rail 3 above.

## For VSCode tasks and terminal windows in VSCode:

The [`EnvironmentVariableCollection`](https://code.visualstudio.com/api/references/vscode-api#ExtensionContext) object is ideal. For existing terminal windows, a little warning icon shows up, asking the user to relaunch the terminal--but this is very easy to miss, so don't count on it always working...

**Reply by bwateratmsft** [reply]
If you find a simpler way though, we're definitely interested. :smile:

**Reply by msorens** [reply]
Ah, you can pass an env override to `spawn` — that is the magic I need.
And yes, this is for the Windows side only; for Unix, I do go straight to `/usr/local/bin`, just as you have.
Your code comments there describe exactly the problem and solution for me. 
Thanks!!

**Reply by msorens** [reply]
One question after reviewing your code: why is the `normalize(combinedEnv)` call required? The comment on`normalize` says "Node defines the environment slightly differently from VS Code", but when I used it I failed to see any differences at all.

**Reply by bwateratmsft** [reply]
I think it depends on how strictly TypeScript is checking things. In a looser mode it's fine but in a stricter mode it wants more precision. I'm guessing it's `strictNullChecks` or something like that?
